### ISSUE TITLE
Unable to create a repository in an organisation

### ISSUE THREAD

**Issue by jeroenjacobs79** [integrations/terraform-provider-github#876]
Hello,

I'm trying to create a repository in an organisation using my personal access token. My config is something like this:

```
provider "github" {
  owner = "headincloud"
}

resource "github_repository" "app_repo" {
  name        = "cluster-apps"
  description = "App repository"

  visibility = "private"
  delete_branch_on_merge = "true"
}

```

I have set the `GITHUB_TOKEN` to my PAT (which I granted all the permissions for now), but the repo still gets created under my own user-name, not onder my organisation.

Now, maybe creating organisation repos doesn't work with PAT's, and I should use OAuth? But I have no clue on how to proceed with this. When trying to create a new GitHub OAuth app, it asks for a homepage URL and callback URL but I have no clue on what I should put here for TerraForm. The documentation is not clear on this either...

So, what is the correct approach here? I believe the documentation could be improved on this subject.

Terraform version: 1.0.3
Github provider version: 4.13.0

**Comment by jeroenjacobs79** [comment]
Update:

`gh repo create headincloud/my-test-repo` works just fine, so PAT's are allowed to create organisation repositories. 

So, the `owner` statement in the `provider` section seems to get ignored?

**Comment by nicksantamaria** [comment]
I've encountered this same problem. Using the `owner` parameter in the `provider` block did not work, but setting the `GITHUB_OWNER` envvar did.

**Comment by glennpratt** [comment]
Seeing the same issue, which appears to be a regression from previous versions

**Comment by serain** [comment]
I faced this issue only when trying to create a repo in a child module. It doesn't pick up the org configuration from the parent module and you have to explicitly pass an aliased provider to the module.

**Comment by mariodavid** [comment]
hi @serain ,

I'm facing the same issue within the module. Can you show an example of how you passed in the provider to the module? Just via a dedicated variable in the module or via the provider meta argument from the module (https://www.terraform.io/docs/language/meta-arguments/module-providers.html)?

**Comment by Sinclert** [comment]
Same issue with current version (`4.17.0`). It seems the documentation block talking about the `owner` / `token` GitHub provider arguments, and how they could be replaced by the `GITHUB_OWNER` and `GITHUB_TOKEN` env. variables **is misleading**.

❌  - Does not work:
```hcl
provider "github" {
  owner = "my-org"
  token = "my-token"
}
```
✅  - Does work:

```hcl
provider "github" {}
```

```
export GITHUB_OWNER="my-org"
```

**Comment by Sinclert** [comment]
Same issue with current version (`4.17.0`).

It seems [the documentation](https://registry.terraform.io/providers/integrations/github/4.17.0/docs) covering the usage of the `owner` / `token` GitHub provider arguments, and how they could be replaced by the `GITHUB_OWNER` and `GITHUB_TOKEN` env. variables **is misleading** 🙅🏻‍♂️ 

In order to get it to work:

- On the provider file:
    ```hcl
    provider "github" {}
    ```

- On the environment running `terraform apply`:
    ```
    export GITHUB_OWNER="your-org-name-here"
    export GITHUB_TOKEN="your-secret-token-here"
   ```

**Comment by kfcampbell** [comment]
I've spent some time playing with this recently, and I'm not able to reproduce the issue. Here's a super pared-down template as an example: 

```hcl
terraform {
  required_providers {
    github = {
      source = "integrations/github"
      version = "4.18.0"
    }
  }
}

provider "github" {
  owner = "kfcampbell-terraform-provider"
  token = "ghp_personal_token_redacted"
}


resource "github_repository" "app_repo" {
  name        = "876-repro"
  description = "App repository"

  visibility = "private"
  delete_branch_on_merge = "true"
}

```

That repo has been created successfully [here](https://github.com/kfcampbell-terraform-provider/876-repro) in my test organization rather than my personal profile. Can you try that reproduction and let me know if it works for you?

**Comment by Sinclert** [comment]
Hi @kfcampbell , thanks for your reply 👋🏻 

I just ran your example, changing the owner between my personal account, and an owned organization account, getting it to work in both. Then I tried switching my `integrations/github` provider back to version `4.17.0` to try to replicate my original setup. However, **I could not get the same results as I got when I wrote in this issue** 🤔 

Perhaps I was confused with another `404` error I got back then (regarding the creation of a branch using a non-existing `main` branch as the default source), and I thought that `404` error was that Terraform was not able to locate the repository in the first place...

Not sure, but it _seems_ solved.

**Comment by akloss-cibo** [comment]
I just encountered this with github provider 4.19.2, terraform 1.1.4.

**Comment by Jonas-Sander** [comment]
This problem occured for me with 4.20.1, settings the `GITHUB_OWNER` environment variable fixed it for me.

**Comment by pksunkara** [comment]
> I faced this issue only when trying to create a repo in a child module. It doesn't pick up the org configuration from the parent module and you have to explicitly pass an aliased provider to the module.

FWIW, this occurs only in the above situation.

**Comment by RulerOf** [comment]
Had this problem, scratched my head on it for an hour or two, but @serain pointed me in the right direction.

The child module didn't have a `required_providers {}` block, so when `terraform init` performed module resolution, it would configure the `integrations/github` provider in the root module, and then bring up `hashicorp/github` in the child module.

The child module of course wouldn't get the provider config because the providers don't match, leading to it thinking that the owner config was blank.

To fix it, in your root module:
```hcl
terraform {
  required_providers {
    github = {
      source  = "integrations/github"
    }
  }
}

provider "github" {
  owner = "YourOrgName" 
  token = "YourToken" // Or use ENV
}
```

Then in the child module:
```hcl
terraform {
  required_providers {
    github = {
      source  = "integrations/github"
    }
  }
}
```

**Comment by kfcampbell** [comment]
@RulerOf this is something that's bitten multiple people. Would you (or anybody else) have any interest in creating a PR for our documentation that might explain the issue?

**Comment by RulerOf** [comment]
@kfcampbell I could certainly take a stab at it, you can assign this issue to me if you like. 

Out of curiosity, this smells like an issue that should be resolved in core terraform. When terraform tries to guess providers based on resources, it should prioritize providers that are explicitly or implicitly passed from parent modules, and not... whatever it's doing presently.

**Comment by kfcampbell** [comment]
Agreed...I used to think (maybe still should think?) I had a misunderstanding of Terraform and now I'm super careful to add a `required_providers` block everywhere. 

This error is probably related to the transition of the provider from Hashicorp- to GitHub-owned.

**Comment by kfcampbell** [comment]
Thanks for giving it a shot! Please @ me on the PR if/when you get to it.

**Comment by rajansingh277** [comment]
Initializing provider plugins...
- Finding integrations/github versions matching "~> 5.0"...
╷
│ Error: Failed to install provider

**Comment by RobbieMcKinstry** [comment]
Bump.

**Comment by RobbieMcKinstry** [comment]
Bump.

**Comment by Bezilon** [comment]
Bump.

**Comment by yudhiesh-oc** [comment]
Still seeing this issue

**Comment by deiga** [comment]
@yudhiesh-oc @Bezilon @RobbieMcKinstry Could you share details about what exactly your configuration is and what is the issue you are facing?

**Comment by RobbieMcKinstry** [comment]
Hi @deiga I'm afraid not. This issue has been open since 2021. It's been awhile since I produced my initially broken code. However, I'd appreciate it if the bot that automatically closes valid issues wouldn't try to close this issue until we have a proper reproduction and remediation.

**Comment by rogermasferrer** [comment]
I just eoncountered this issue and I solved it adding the required_providers block into any terraform module that uses github resources. With that I didn't need the GITHUB_OWNER variable set anymore to create repos in my org, like this:

`terraform {
...
  required_providers {
    github = {
      source  = "integrations/github"
      version = "~> 6.0"
    }
  }
...
}`

**Comment by deiga** [comment]
Without any logs and examples from a more recent version of the provider we have to assume that this issue is resolved and are closing this

**Comment by RulerOf** [comment]
It is possible to reproduce this, but it's typically always a configuration problem.

If this affects you: ensure that the `integrations/github` provider is configured in the `terraform { required_providers {} }` block in your root module and in every submodule down to where your resources are defined. 

For a concrete example, see [my comment](https://github.com/integrations/terraform-provider-github/issues/876#issuecomment-1303790559) further up the thread.

----

@deiga I humbly suggest locking this discussion to ensure this comment remains visible into the future.

**Comment by volker-raschek** [comment]
Hi @deiga, 
I've create a GitHub repository to reproduce the issue: https://github.com/volker-raschek/terraform-github-provider-issue

Please go through the README file to reproduce it.

**Comment by deiga** [comment]
> Hi @deiga, 
> I've create a GitHub repository to reproduce the issue: https://github.com/volker-raschek/terraform-github-provider-issue
> 
> Please go through the README file to reproduce it.
> 

Thanks for that! Amazing!

Without running it, I can see that the module doesn't have the required_providers config, which is my main guess at the issue. 

This is not something we can fix AFAIK, as long as hashicorp/github provider exists

**Comment by volker-raschek** [comment]
Hi @deiga, 
perhaps I am not yet as familiar with Terraform as some others, but the provider `integrations/github` is defined in `required_providers` of the `terraform.tf` [file](https://github.com/volker-raschek/terraform-github-provider-issue/blob/1a32d1d841fb36ee931259564e4ea43cec5da35d/terraform.tf#L2). 

Should be the provider defined twice, in the root dir and submodule `github_repository`?

**Comment by RulerOf** [comment]
> Should be the provider defined twice, in the root dir and submodule github_repository?

Yes @volker-raschek. [The example above](https://github.com/integrations/terraform-provider-github/issues/876#issuecomment-1303790559) makes it clear. You must add:

```hcl
terraform {
  required_providers {
    github = {
      source  = "integrations/github"
    }
  }
}
```

To your child module somewhere.

**Comment by deiga** [comment]
You need to consider that a module is a semi-autonomous part, which knows only what's given into it or defined inside it. Maybe like a Class in OOP. 
And this extends to dependencies in a limited fashion, but due to the naming confusion of the legacy hashicorp/github provider (and auto-resolution of Hashicorp providers) it creates this silly situation.

**Comment by RulerOf** [comment]
@deiga it does stand to reason that a patch to core terraform is called for by this issue. Although I've got no idea what their appetite would be to change this kind of default.

**Comment by deiga** [comment]
@RulerOf Feel free to file an Issue with them, I don't have my hopes up 😬
